cpp/include/raft/cluster/kmeans_balanced.cuh

```diff
@@ -0,0 +1,366 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <raft/cluster/detail/kmeans_balanced.cuh>
+#include <raft/core/mdarray.hpp>
+#include <raft/util/cuda_utils.cuh>
+
+namespace raft::cluster::kmeans_balanced {
+
+/**
+ * @brief Find clusters of balanced sizes with a hierarchical k-means algorithm.
+ *
+ * This variant of the k-means algorithm first clusters the dataset in mesoclusters, then clusters
+ * the subsets associated to each mesocluster into fine clusters, and finally runs a few k-means
+ * iterations over the whole dataset and with all the centroids to obtain the final clusters.
+ *
+ * Each k-means iteration applies expectation-maximization-balancing:
+ *  - Balancing: adjust centers for clusters that have a small number of entries. If the size of a
+ *    cluster is below a threshold, the center is moved towards a bigger cluster.
+ *  - Expectation: predict the labels (i.e find closest cluster centroid to each point)
+ *  - Maximization: calculate optimal centroids (i.e find the center of gravity of each cluster)
+ *
+ * The number of mesoclusters is chosen by rounding the square root of the number of clusters. E.g
+ * for 512 clusters, we would have 23 mesoclusters. The number of fine clusters per mesocluster is
+ * chosen proportionally to the number of points in each mesocluster.
+ *
+ * This variant of k-means uses random initialization and a fixed number of iterations, though
+ * iterations can be repeated if the balancing step moved the centroids.
+ *
+ * Additionally, this algorithm supports quantized datasets in arbitrary types but the core part of
+ * the algorithm will work with a floating-point type, hence a conversion function can be provided
+ * to map the data type to the math type.
+ *
+ * @code{.cpp}
+ *   #include <raft/core/handle.hpp>
+ *   #include <raft/cluster/kmeans_balanced.cuh>
+ *   #include <raft/cluster/kmeans_balanced_types.hpp>
+ *   ...
+ *   raft::handle_t handle;
+ *   raft::cluster::kmeans_balanced_params params;
+ *   auto centroids = raft::make_device_matrix<float, int>(handle, n_clusters, n_features);
+ *   raft::cluster::kmeans_balanced::fit(handle, params, X, centroids.view());
+ * @endcode
+ *
+ * @tparam DataT Type of the input data.
+ * @tparam MathT Type of the centroids and mapped data.
+ * @tparam IndexT Type used for indexing.
+ * @tparam MappingOpT Type of the mapping function.
+ * @param[in]  handle     The raft handle
+ * @param[in]  params     Structure containing the hyper-parameters
+ * @param[in]  X          Training instances to cluster. The data must be in row-major format.
+ *                        [dim = n_samples x n_features]
+ * @param[out] centroids  The generated centroids [dim = n_clusters x n_features]
+ * @param[in]  mapping_op (optional) Functor to convert from the input datatype to the arithmetic
```

Comment:
Is there a technical requirement to have the identity mapping when `DataT` == `MathT`? At the first glance, the mapping_op sounds like a useful feature even in the case when the types match. If you have a static check for the equality of the types somewhere, perhaps you can replace/complement it with comparing the `MappingOpT` to `identity_op` (static as well).

Reply:
There are a few reasons to have this assumption:

- We don't want to have to copy the dataset when unnecessary + affect the batch size.
- There are functions that can take the data transformed or not yet. The way to know whether it has been transformed already at the moment is the type of data. I ran into an issue earlier where I was transforming it twice. Note that this wasn't an issue with the ANN `mapping<float>` precisely because it behaves like an identity if the type is already float.

I'm not saying it impossible, just unnecessarily complex and not presently needed.

Reply:
I don't see exactly where is it unnecessarily complex? This just sounds like a rather artificial constraint.

Reply:
Unnecessary because we don't have any use for this yet; the other k-means algorithm doesn't support it either. And though it's not particularly hard to implement, it requires some changes and additional checks, and proper testing too, and I don't want to hold this PR for this.

I think that we can always do this change later. As the identity is currently assumed when the type is the same, it's fair to consider it non-breaking to change that behavior in the future (a user should really not be passing a non-identity op if the type is the same).

Reply:
Ok, nevermind.